Source/WTF/wtf/darwin/TypeCastsOSObject.h

```diff
@@ -0,0 +1,224 @@
+
+#define WTF_OS_OBJECT_DISPATCH_TYPES(M) \
+    M(dispatch_group) \
+    M(dispatch_object) \
+    M(dispatch_queue) \
+    M(dispatch_queue_global) \
+    M(dispatch_source)
+
+WTF_EXTERN_C_BEGIN
+#define WTF_DECLARE_OS_OBJECT_DISPATCH_BASE_STRUCT(TypeName) \
+struct TypeName##_s;
+WTF_OS_OBJECT_DISPATCH_TYPES(WTF_DECLARE_OS_OBJECT_DISPATCH_BASE_STRUCT)
+#undef WTF_DECLARE_OS_OBJECT_DISPATCH_BASE_STRUCT
+WTF_EXTERN_C_END
+
+#define WTF_OS_OBJECT_NETWORK_TYPES(M) \
+    M(nw_endpoint) \
+    M(nw_resolution_report) \
+    M(nw_path)
+
+WTF_EXTERN_C_BEGIN
+#define WTF_DECLARE_OS_OBJECT_NETWORK_BASE_STRUCT(TypeName) \
+struct TypeName;
+WTF_OS_OBJECT_NETWORK_TYPES(WTF_DECLARE_OS_OBJECT_NETWORK_BASE_STRUCT)
+#undef WTF_DECLARE_OS_OBJECT_NETWORK_BASE_STRUCT
+WTF_EXTERN_C_END
+
+#define WTF_OS_OBJECT_TYPES(M) \
+    WTF_OS_OBJECT_DISPATCH_TYPES(M) \
+    WTF_OS_OBJECT_NETWORK_TYPES(M)
+
+// Because ARC enablement is a compile-time choice, and we compile this header
+// both ways, we need a separate copy of our code when ARC is enabled.
+#if __has_feature(objc_arc)
+#define checked_osobject_cast checked_osobject_castARC
+#define dynamic_osobject_cast dynamic_osobject_castARC
+#endif
+
+namespace WTF {
+
+template<typename> struct OSObjectTypeCastTraits;
+#define WTF_DECLARE_OS_OBJECT_TYPE_CAST_TRAITS_INTERNAL(TypeName, Suffix) \
+template<> struct OSObjectTypeCastTraits<TypeName##_t> { \
+    using BaseType = struct TypeName##Suffix*; \
+};
+
+#define WTF_DECLARE_OS_OBJECT_DISPATCH_TYPE_CAST_TRAITS(TypeName) \
+WTF_DECLARE_OS_OBJECT_TYPE_CAST_TRAITS_INTERNAL(TypeName, _s)
+WTF_OS_OBJECT_DISPATCH_TYPES(WTF_DECLARE_OS_OBJECT_DISPATCH_TYPE_CAST_TRAITS)
+#undef WTF_DECLARE_OS_OBJECT_DISPATCH_TYPE_CAST_TRAITS
+
+#define WTF_DECLARE_OS_OBJECT_NETWORK_TYPE_CAST_TRAITS(TypeName) \
+WTF_DECLARE_OS_OBJECT_TYPE_CAST_TRAITS_INTERNAL(TypeName, )
+WTF_OS_OBJECT_NETWORK_TYPES(WTF_DECLARE_OS_OBJECT_NETWORK_TYPE_CAST_TRAITS)
+#undef WTF_DECLARE_OS_OBJECT_NETWORK_TYPE_CAST_TRAITS
+
+#undef WTF_DECLARE_OS_OBJECT_TYPE_CAST_TRAITS_INTERNAL
+
+// Must define this for is_osObject<T>() when not building with Objective-C.
+#ifndef OS_OBJECT_CLASS
+#define OS_OBJECT_CLASS(name) OS_##name
+#endif
+
+template<typename T> bool is_osObject(CFTypeRef);
+
+#define WTF_IMPLEMENT_IS_OS_OBJECT_FUNCTIONS(TypeName) WTF_IMPLEMENT_IS_OS_OBJECT_FUNCTIONS_INTERNAL(TypeName##_t, STRINGIZE_VALUE_OF(OS_OBJECT_CLASS(TypeName)))
+#define WTF_IMPLEMENT_IS_OS_OBJECT_FUNCTIONS_INTERNAL(TypeName, ProtocolString) \
+template<> inline bool is_osObject<OSObjectTypeCastTraits<TypeName>::BaseType>(CFTypeRef object) \
+{ \
+    Class cls = object_getClass(bridge_id_cast(object)); \
+    return class_conformsToProtocol(cls, objc_getProtocol(ProtocolString)); \
+} \
+
+WTF_OS_OBJECT_TYPES(WTF_IMPLEMENT_IS_OS_OBJECT_FUNCTIONS)
+#undef WTF_IMPLEMENT_IS_OS_OBJECT_FUNCTIONS_INTERNAL
+#undef WTF_IMPLEMENT_IS_OS_OBJECT_FUNCTIONS
+
+#ifdef __OBJC__
+
+template<typename T> inline bool is_osObject(id object)
```

Comment:
I think we want to use camelCase. There is a bug / PR to update retain ptr functions: https://github.com/WebKit/WebKit/pull/45093

Reply:
Okay, I'll change it to `isOSObject`.